Assets/MixedRealityToolkit.Extensions/Socketer/Scripts/Socketer.cs

```diff
@@ -0,0 +1,173 @@
+using Windows.Foundation;
+using Windows.Networking;
+using Windows.Storage.Streams;
+using Windows.Networking.Connectivity;
+using System.Threading;
+#endif
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.Socketer
+{
+#if UNITY_EDITOR || UNITY_STANDALONE || UNITY_WSA || UNITY_IOS
+using UnityEngine;
+
+    /// <summary>
+    /// Sends and receives packets over the network, using the chosen protocol.
+    /// </summary>
+    public class Socketer : MonoBehaviour
+    {
+        /// <summary>
+        /// The protocol to use:  TCP or UDP.
+        /// </summary>
+        public SocketerClient.Protocol Protocol;
+        /// <summary>
+        /// Whether this object should listen for incoming connections/packets, or send them.
+        /// TCP Socketers are able to both send and recieve once a connection is intitiated.
+        /// </summary>
+        public SocketerClient.ProtocolDirection Direction;
+        /// <summary>
+        /// The IP address to connect to, if the Direction is Sender.  Ignored if Listener.
+        /// </summary>
+        public string Host;
+        /// <summary>
+        /// The port to use.  Suggested values are between 10000 and 40000, and must agree
+        /// with the Socketer on the other end.
+        /// </summary>
+        public int Port;
+
+        /// <summary>
+        /// Fired when a message arrives.  Runs on the Update thread.
+        /// </summary>
+        public event Action<Socketer, MessageEvent> Message;
+        /// <summary>
+        /// Fired when a TCP connection completes a connection.  The string is the Host of the
+        /// other Socketer.
+        /// Note:  this may fire on a thread other than the Update thread.
+        /// </summary>
+        public event Action<Socketer, int, string> Connected;
+        /// <summary>
+        /// Fired when a TCP connection loses a connection.  The string is the Host of the
+        /// other Socketer.
+        /// Note:  this may fire on a thread other than the Update thread.
+        /// </summary>
+        public event Action<Socketer, int, string> Disconnected;
+
+        private SocketerClient socketer;
+        private Queue<MessageEvent> stateQueue = new Queue<MessageEvent>();
+
+        internal const UInt32 HandshakeCode = 9999;
+
+        /// <summary>
+        /// Gets the IP of this machine.  Suitable for use as a Host on another Socketer.
+        /// </summary>
+        /// <returns></returns>
+        public static string GetLocalIPAddress()
+        {
+            return SocketerClient.GetLocalIPAddress();
+        }
+
+        /// <summary>
+        /// Sends a byte array to the other side.  Works for all Socketers except UDP Listeners.
+        /// If the other Socketer is not listening, the messages are lost.
+        /// </summary>
+        /// <param name="message">Message contents.</param>
+        public void SendNetworkMessage(byte[] message)
+        {
+            socketer.SendNetworkMessage(message);
```

Comment:
Since this is a public method, it can be called even if Socketer is disabled, which means that it's prone to throwing null pointer exceptions